Add observability getting start doc (#2290)
giantswarm/docs: src/content/getting-started/observe-your-clusters-and-apps/_index.md

---
title: Observe your clusters and apps

> description: Check cluster and app metrics with the observability tools provided with the Giant Swarm platform.

Reviewer: Is there a reason why you decided to remove logs here?

Reviewer: sure :)

> description: Check cluster and app metrics with the observability tools provided with the Giant Swarm platform.

Reviewer: @Rotfuks would you take another look?

Also, consider that ingesting new metrics into the platform impacts your costs. The central monitoring system's resource consumption is actually related to the number of metrics it has to handle, so please choose which metrics you want to ingest.

## Step 1: Create a service monitor

In case of running the `hello-world` application, there is a baked-in service monitor that scrapes the metrics from the application. You can enable it by setting `serviceMonitor.enabled` to `true` in the values of the helm chart. If you are running your own application, you need to create a service monitor that scrapes the metrics from your application. This is an example of a service monitor you can use:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    application.giantswarm.io/team: team-rocket
    app.kubernetes.io/instance: my-service
  name: my-service
  namespace: my-namespace
spec:
  endpoints:
  - interval: 60s
    path: /metrics
    port: web
  selector:
    matchLabels:
      app.kubernetes.io/instance: my-service
```

> An important bit to notice is the `application.giantswarm.io/team` label. This label is necessary as it's required for the `Prometheus` agent to discover the target. At the same time, it's used as unique identifier for your tenant or team.

Reviewer: ```suggestion
An important bit to notice is the `application.giantswarm.io/team` label. This label is necessary as it's required for the `Prometheus` agent to discover the target.
```

This might change, it might be better to not say it?

Reading the manifest, the application will be scraped every 60 seconds. It will read the metrics from the `/metrics` endpoint using the [port name](https://kubernetes.io/docs/concepts/services-networking/service/#field-spec-ports) as a regular service does. Also, the `app.kubernetes.io/instance` label is used to identify the application, so it should match the application label.

After applying the service monitor, you can open the `Explore` view in the `Grafana` UI and start querying the metrics from your application.

![Explore application metrics](explore-application-metrics.png)

> __Note__: The `Grafana` UI is accessible via ingress. To know the address you can check the ingress resource in the `monitoring` namespace of platform API.

Author: ```suggestion
__Note__: The `Grafana` UI is accessible via ingress. To find the address, you can check the ingress resource in the `monitoring` namespace of the platform API. More information in the [access `Grafana` page]({{< relref "/tutorials/observability/data-exploration/accessing-grafana/" >}})
```

> __Note__: The `Grafana` UI is accessible via ingress. To know the address you can check the ingress resource in the `monitoring` namespace of platform API.

Reviewer: Is this really how we want customers to get the ingress?

Author: well it is the way we have now, with Backstage will be easy but it is not there yet

Reviewer: I think we have a page explaining how to access grafana. Maybe we can link it here instead?

## Step 2: Inspect the default dashboards

The platform provides a set of default dashboards that you can use to monitor the cluster components and your application. You can access them by opening the `Dashboards` view in the `Grafana` UI and explore different options in `Giant Swarm Public Dashboards` folder.

Below is an example of service monitor metrics filtered to show the `hello-world` application and how many cardinality the `promhttp_metric_handler_requests_total` metric has.

![Service monitor metrics](service-monitor-metrics.png)

Along with service monitor metrics, you can check metrics for DNS, Ingress, Flux, and many other components. Don't hesitate to provide feedback on the dashboards or suggest new ones.

## Step 3: Create your own dashboard

In some occasions, you may want to create a custom dashboard to visualize the metrics of your application. Once logged into the [platform API]({{< relref "/vintage/use-the-api/management-api/_index.md" >}}) you can create a special `ConfigMap` containing the dashboard in JSON format. It will look like this:

```yaml

> apiVersion: v1

Reviewer: This duplicates the other doc right? I'm a bit skeptical because then we will need to change it in multiple places?

Author: I would like to move it here and, in the tutorial, add something more advanced. WDYT?

Reviewer: Sure :)

data:
  my-dashboard.json: |-
    __DASHBOARD_JSON__
kind: ConfigMap
metadata:
  annotations:

>     k8s-sidecar-target-directory: /var/lib/grafana/dashboards/giant-swarm-public-dashboards

Reviewer: ```suggestion
    k8s-sidecar-target-directory: /var/lib/grafana/dashboards/customer
```

  labels:
    app.giantswarm.io/kind: dashboard
  name: my-grafana-dashboard
  namespace: my-namespace
```

It's important to have an annotation with the key `k8s-sidecar-target-directory` pointing to the location where the dashboard will be stored in the `Grafana UI`. Also notice the label `app.giantswarm.io/kind` has to be set to `dashboard` in order to be recognized by the platform.

> In our example you can find the dashboard JSON [here](./dashboard.json). Download the dashboard content and use `jq` to minify it:

Author: ```suggestion
The easiest way to build a dashboard is using the Grafana UI. There, you can create panels with the desired visualizations and then export them by selecting `Share > Export`` in the dashboard context menu or by accessing the JSON Model in the dashboard settings.

__Warning__: The dashboards served in the platform are all created from a GitOps repository or created by users through config maps; there is no persistent storage for `Grafana`, which means dashboards are not persisted if the system restarts or are upgraded.

In our example, you can find the dashboard JSON [here](./dashboard.json). Download the dashboard content and use `jq` to minify it:
```

```bash
jq -c . dashboard.json
```

Then replace the `__DASHBOARD_JSON__` placeholder in the `ConfigMap` with the minified JSON content. After that, apply the `ConfigMap` to the platform API:

```bash
kubectl apply -f config-map.yaml
```

> After few seconds, you can open the `Dashboards` view in the `Grafana` UI and find your custom dashboard in the `Giant Swarm Public Dashboards` folder.

Author: ```suggestion
After few seconds, you can open the `Dashboards` view in the `Grafana` UI and find your custom dashboard in the `Customer` folder.
```

![custom-dashboard](custom-dashboard.png)

> As you can see in the image above, the dashboard shows some information about the `hello-world` application. You see the number of errors, the number of success requests, and number of requests by code. Feel free to poke around and create your own panels.

Author: ```suggestion
As you can see in the image above, the dashboard shows some information about the `hello-world` application. You see the number of errors, the number of success requests, and the number of requests by code. Feel free to poke around and create your own panels.
```

title: Observe your clusters and apps
description: Check cluster and app metrics and logs with the observability tools provided with the Giant Swarm platform.
weight: 70
last_review_date: 2024-09-25
owner:
  - https://github.com/orgs/giantswarm/teams/sig-docs
user_questions:
  - How do I observe the platform metrics and logs for my application?
  - What do I need to do to observe the platform metrics and logs for my application?
---

> <Explanation how monitoring works in two paragraphs>

Author: @QuentinBisson, it would be great if you are able to add an intro here; you can write it using the suggestion function. I take the rest of the page so I practice it

Reviewer: I think this might cover it. @Rotfuks as you have better writing skills :D 

```suggestion
Each cluster created on the Giant Swarm platform benefits from our monitoring feature which allows us to provide you with 24/7 support to ensure best quality of service. The metrics scraped by a in-cluster agent and sent to a central location on the management cluster for you to be able to interact with them through our managed Grafana instance.
```

user_questions:
  - How do I observe the platform metrics and logs for my application?
  - What do I need to do to observe the platform metrics and logs for my application?
---

In the Giant Swarm platform every cluster benefits from automatic monitoring by default. It enables you to observe your application with little effort. An in-cluster agent scrapes the metrics and sends them to a central location on the management cluster, where you can interact with them through our managed Grafana instance.

Learn how to observe your application metrics and logs with the observability tools provided with the Giant Swarm platform.

## Requirements

First step, you need a running workload cluster. If you don't have one, please first [create a workload cluster]({{< relref "/getting-started/provision-your-first-workload-cluster" >}}). Second, you need to deploy the `hello-world` application explained [in previous entry]({{< relref "/getting-started/install-an-application" >}}).

In case you are running your own application, you need to make sure that your application is already [instrumented](https://opentelemetry.io/docs/concepts/instrumentation/) to export metrics.

> Also consider that ingesting new metrics into the platform impacts your costs. The resource consumption of the central monitoring system is related to the amount of metrics it has to handle, so choose wisely which metrics you want to ingest.

Author: ```suggestion
Also consider that ingesting new metrics into the platform impacts your costs. The central monitoring system's resource consumption is related to the number of metrics it has to handle, so choose which metrics you want to ingest carefully.
```
